python_modules/libraries/dagster-dg/dagster_dg/config.py

```diff
@@ -28,9 +41,112 @@ def _get_default_cache_dir() -> Path:
 DEFAULT_CACHE_DIR = _get_default_cache_dir()
 
 
+def discover_workspace_root(path: Path) -> Optional[Path]:
+    workspace_config_path = discover_config_file(
+        path, lambda config: config["directory_type"] == "workspace"
+    )
+    return workspace_config_path.parent if workspace_config_path else None
+
+
+def discover_config_file(
+    path: Path,
+    predicate: Optional[Callable[[Mapping[str, Any]], bool]] = None,
+) -> Optional[Path]:
+    current_path = path.absolute()
+    while True:
+        config_path = current_path / "pyproject.toml"
+        if config_path.exists() and has_dg_file_config(config_path, predicate):
+            return config_path
+        if current_path == current_path.parent:  # root
+            return
+        current_path = current_path.parent
+
+
+# NOTE: The set/has/get_config_from_cli_context is only used for the dynamically generated scaffold
+# component commands. Hopefully we can get rid of it in the future.
+
+_CLI_CONTEXT_CONFIG_KEY = "config"
+
+
+def set_config_on_cli_context(cli_context: click.Context, config: "DgRawCliConfig") -> None:
+    cli_context.ensure_object(dict)
+    cli_context.obj[_CLI_CONTEXT_CONFIG_KEY] = config
+
+
+def has_config_on_cli_context(cli_context: click.Context) -> bool:
+    return _CLI_CONTEXT_CONFIG_KEY in cli_context.ensure_object(dict)
+
+
+def get_config_from_cli_context(cli_context: click.Context) -> "DgRawCliConfig":
+    cli_context.ensure_object(dict)
+    return cli_context.obj[_CLI_CONTEXT_CONFIG_KEY]
+
+
+# ########################
+# ##### MAIN
+# ########################
+
+
 @dataclass
 class DgConfig:
-    """Global configuration for Dg.
+    cli: "DgCliConfig"
```

Comment:
could consider DgGlobalCliConfig?

Reply:
That would disrupt the naming convention in use where classes match the config section names. And I don't think we want `[tool.dg.global_cli]`.